### PR TITLE
Less fragile scheme equality

### LINKED ISSUE #12763: Mangle Names vs. Scheme Equality
#### Description of the problem

The `Mangle Names` option breaks the `Scheme Equality` command.

```coq
Set Mangle Names.
Scheme Equality for nat.
(*
Unexpected error during scheme creation: variable _1 unbound
nat_beq is defined
*)
```

I haven't yet been able to produce the same problem without `Mangle Names`.

#### Coq Version

Coq master 55f4095fe3c366a9f310584a55e2dc0605e5409c (Around July 28, 2020)

### PR DESCRIPTION
**Pull request by jashug**
**Kind:** bug fix

Fixes / closes #12763

We try to use names that are properly fresh.

- [x] Added / updated test-suite

### CODE REVIEW

File: vernac/auto_ind_decl.ml [hunk 1]
```diff
@@ -556,32 +556,36 @@ let list_id l = List.fold_left ( fun a decl -> let s' =
               Id.of_string (s'^"_lb"))
             ::a
     ) [] l
+
+let avoid_of_list_id list_id =
+  List.fold_right (fun (s,seq,sbl,slb) -> List.fold_right Id.Set.add [s;seq;sbl;slb])
```
Comment (SkySkimmer) [inline vernac/auto_ind_decl.ml:561]: Don't fold_right where fold_left works

**Review by herbelin — APPROVED** [review]
This looks good. Thanks a lot.

@SkySkimmer, will you assign?

**Comment by SkySkimmer** [comment]
@herbelin  Why not you assign?

**Comment by herbelin** [comment]
OK, will merge tomorrow.

**Comment by Zimmi48** [comment]
@jashug @herbelin @SkySkimmer: Should we backport this PR? #12881 depends on it, so there's no way to backport it without backporting this one as well. It was not marked for 8.12.1 inclusion but I cannot really see why.

**Comment by herbelin** [comment]
This PR is a big larger but I would say that it is probably ok to backport it to 8.12.1. Otherwise, backporting the changes to `tactics.ml` and the additions to `tactics.mli` would be enough.

**Comment by jashug** [comment]
I don't have a good sense of the criteria for backporting a PR.

* This PR is a bugfix for a bug which is visible even without the still experimental option `Mangle Names`.
* #12881 also includes such a bugfix.
* As herbelin said, the changes to `tactics.ml` and `tactics.mli` are the only depencency between these two PRs.
* I think that the newly introduced `intro_using_then` and `intros_using_then` are useful for writing correct plugins.

**Comment by herbelin** [comment]
Well, if noone else objects, I'd also lean towards backporting #12881 as well.

**Comment by Zimmi48** [comment]
Yes, I agree. Let's backport the two PRs (if no opposition from @ejgallego).
However, as I said in #12881, I won't backport the deprecation of `intros_using` in favor of `intros_using_then` since the latter didn't exist in 8.12.

**Pull request merged**